stateproof/builder.go

```diff
@@ -20,48 +20,163 @@ import (
+		}
+	}
+}
+
+// loadOrCreateBuilderWithSignatures either loads a builder from the DB or creates a new builder.
+// this function fills the builder with all the available signatures
+func (spw *Worker) loadOrCreateBuilderWithSignatures(rnd basics.Round) (builder, error) {
+	b, err := spw.loadOrCreateBuilder(rnd)
+	if err != nil {
+		return builder{}, err
+	}
+
+	if err := spw.loadSignaturesIntoBuilder(&b); err != nil {
+		return builder{}, err
+	}
+	return b, nil
+}
+
+func (spw *Worker) loadOrCreateBuilder(rnd basics.Round) (builder, error) {
+	var buildr builder
+	err := spw.db.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+		buildr, err = getBuilder(tx, rnd)
+		return err
+	})
+
+	if err == nil {
+		return buildr, nil
+	}
+
+	if !errors.Is(err, sql.ErrNoRows) {
+		spw.log.Errorf("loadOrCreateBuilder: error while fetching builder from DB: %v", err)
+	}
+
+	buildr, err = createBuilder(rnd, spw.ledger)
+	if err != nil {
+		return builder{}, err
+	}
+
+	err = spw.db.Atomic(func(_ context.Context, tx *sql.Tx) error {
+		return persistBuilder(tx, rnd, &buildr)
+	})
+
+	// We ignore persisting errors because we still want to try and use our successfully generated builder,
+	// even if, for some reason, persisting it failed.
+	if err != nil {
+		spw.log.Errorf("loadOrCreateBuilder(%d): failed to insert builder into database: %v", rnd, err)
+	}
+
+	return buildr, nil
+}
+
+func (spw *Worker) loadSignaturesIntoBuilder(buildr *builder) error {
+	var sigs []pendingSig
+	err := spw.db.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		var err2 error
+		sigs, err2 = getPendingSigsForRound(tx, basics.Round(buildr.Round))
+		return err2
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, sig := range sigs {
+		err = buildr.insertSig(&sig, false)
+		if err != nil {
+			spw.log.Warn(err)
+		}
+	}
+	return nil
+}
+
+func createBuilder(rnd basics.Round, votersFetcher ledgercore.LedgerForSPBuilder) (builder, error) {
+	// since this function might be invoked under tracker commit context (i.e invoked from the ledger code ),
+	// it is important that we do not use the ledger directly.
+
+	hdr, err := votersFetcher.BlockHdr(rnd)
 	if err != nil {
 		return builder{}, err
 	}
 
 	hdrProto := config.Consensus[hdr.CurrentProtocol]
 	votersRnd := rnd.SubSaturate(basics.Round(hdrProto.StateProofInterval))
-	votersHdr, err := l.BlockHdr(votersRnd)
+	votersHdr, err := votersFetcher.BlockHdr(votersRnd)
 	if err != nil {
 		return builder{}, err
 	}
 
 	lookback := votersRnd.SubSaturate(basics.Round(hdrProto.StateProofVotersLookback))
-	voters, err := l.VotersForStateProof(lookback)
+	voters, err := votersFetcher.VotersForStateProof(lookback)
 	if err != nil {
 		return builder{}, err
 	}
 
 	if voters == nil {
 		// Voters not tracked for that round.  Might not be a valid
 		// state proof round; state proofs might not be enabled; etc.
-		return builder{}, fmt.Errorf("voters not tracked for lookback round %d", lookback)
```

Comment:
Every network will trigger this line once. For new networks (e2e tests): this will get triggered when the builder tries to create state proof for round 256, and will try to lookup voters for round 0.

Reply:
you can see this warning message in `TestStateProofRecovery` test.